libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+		return (enum speaker_layout) channels;
+	return SPEAKERS_UNKNOWN;
+}
+
+static enum audio_format pulseaudio_to_obs_audio_format(
+	pa_sample_format_t format)
+{
+	switch (format) {
+	case PA_SAMPLE_U8:
+		return AUDIO_FORMAT_U8BIT;
+	case PA_SAMPLE_S16LE:
+		return AUDIO_FORMAT_16BIT;
+	case PA_SAMPLE_S32LE:
+		return AUDIO_FORMAT_32BIT;
+	case PA_SAMPLE_FLOAT32LE:
+		return AUDIO_FORMAT_FLOAT;
+	default:
+		return AUDIO_FORMAT_UNKNOWN;
+	}
+}
+
+static void process_byte(void *p, size_t frames, size_t channels, float vol)
+{
+	register char *cur = (char *) p;
+	register char *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+static void process_short(void *p, size_t frames, size_t channels, float vol)
+{
+	register short *cur = (short *) p;
+	register short *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+static void process_float(void *p, size_t frames, size_t channels, float vol)
+{
+	register float *cur = (float *) p;
+	register float *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+void process_volume(const struct audio_monitor *monitor, float vol,
+                    uint8_t *const *resample_data, uint32_t resample_frames)
+{
+	switch (monitor->bytes_per_channel) {
+	case 1:
+		process_byte(resample_data[0], resample_frames,
+		             monitor->channels, vol);
+		break;
+	case 2:
+		process_short(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	default:
+		process_float(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	}
+}
+
+static void do_stream_write(void *param)
+{
+	PULSE_DATA(param);
+	uint8_t *buffer = NULL;
+
+	while (data->new_data.size >= data->buffer_size &&
+	       data->bytesRemaining > 0) {
+		size_t bytesToFill = data->buffer_size;
+
+		if (bytesToFill > data->bytesRemaining)
+			bytesToFill = data->bytesRemaining;
+
+		pa_stream_begin_write(data->stream, (void **) &buffer,
+		                                &bytesToFill);
+
+		circlebuf_pop_front(&data->new_data, buffer, bytesToFill);
+
+		pulseaudio_lock();
+		pa_stream_write(data->stream, buffer, bytesToFill, NULL,
+		                      0LL,
+		                      PA_SEEK_RELATIVE);
+		pulseaudio_unlock();
+
+		data->bytesRemaining -= bytesToFill;
+	}
+}
+
+static void on_audio_playback(void *param, obs_source_t *source,
+                              const struct audio_data *audio_data, bool muted)
+{
+	struct audio_monitor *monitor = param;
+	float vol = source->user_volume;
+	size_t bytes;
+
+	uint8_t *resample_data[MAX_AV_PLANES];
+	uint32_t resample_frames;
+	uint64_t ts_offset;
+	bool success;
+
+	if (pthread_mutex_trylock(&monitor->playback_mutex) != 0) {
+		return;
+	}
+	if (os_atomic_load_long(&source->activate_refs) == 0) {
+		goto unlock;
+	}
+
+	success = audio_resampler_resample(monitor->resampler, resample_data,
+	                                   &resample_frames, &ts_offset,
```

Comment:
Style: Indent with two tabs for continuing a line.  See also the other continuations of this line.